### PR TITLE
Pipelines Refactor - Initial Impl

### PR DESCRIPTION
**Pull request by bfineran**
Initial implementation of pipelines refactor, description will be updated with motivation and framework details before landing

**test_plan:**
simple E2E test included

### CODE REVIEW

File: src/deepsparse/v2/operators/operator.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        """
+        Parses inputs to this Operator and runs the run() method of this operator
+
+        :param args: an unnamed arg may only be provided
+            if it is of the type of the input_schema
+        :param context: pipeline context to pass to operator
+        :param kwargs: kwargs when not initializing from an instantiated schema
+        :return: operator output
+        """
+        if len(args) > 1:
```
Comment (markurtz) [inline src/deepsparse/v2/operators/operator.py:73]: let's see if we can push this to kwargs only for inputs to simplify understanding

File: src/deepsparse/v2/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.validate()
+
+        # SchedulerGroup handles running all schedulers in order of priority
+        self._scheduler_group = SchedulerGroup(self.schedulers)
+
+    def __call__(self, *args, return_context: bool = False, **kwargs):
```
Comment (markurtz) [inline src/deepsparse/v2/pipeline.py:63]: Need to have three cases here:
- sync
- async
- generator

File: src/deepsparse/v2/operators/operator.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    """
+    Base operator class - can represent any part of an ML pipeline
+    """
+
+    # expected structured input and output types, to be defined by child classes
+    input_schema: Optional[Type[OperatorSchema]] = None
+    output_schema: Optional[Type[OperatorSchema]] = None
+
+    @abstractmethod
+    def run(self, inp: OperatorSchema, context: Context) -> OperatorSchema:
```
Comment (dsikka) [inline src/deepsparse/v2/operators/operator.py:36]: If the input_schema isn't mandatory, the typing on this shouldn't always be OperatorSchema?

File: src/deepsparse/v2/operators/operator.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+from pydantic import BaseModel
+
+from deepsparse.v2.utils import Context, OperatorSchema
+
+
+__all__ = ["Operator"]
+
+
+class Operator(ABC):
```
Comment (dbogunowicz) [inline src/deepsparse/v2/operators/operator.py:26]: Are we also going to define an abstract class for EngineOperators (mentioned in your design doc some time ago)?

File: src/deepsparse/v2/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+from deepsparse.v2.operators import Operator
+from deepsparse.v2.routers import Router
+from deepsparse.v2.schedulers import OperatorScheduler, SchedulerGroup
+
+
+__all__ = ["Pipeline"]
+
+
+class Pipeline(BaseModel):
```
[the remark this answers is not in the document]
Comment (dbogunowicz) [reply src/deepsparse/v2/pipeline.py:28]: Why not? Why are you worried about it Dipika?

**Comment by bfineran** [comment]
merging per request by @dsikka, requested changes to come in follow up PRs

**Pull request merged**